pyvista/plotting/_plotting.py

```diff
@@ -0,0 +1,147 @@
+"""These are private methods we keep out of plotting.py to simplfy the module."""
+import warnings
+
+import numpy as np
+import pyvista
+
+from pyvista.utilities import get_array, is_pyvista_dataset
+from .tools import opacity_transfer_function
+
+
+def _has_matplotlib():
+    try:
+        import matplotlib
+        return True
+    except ImportError:  # pragma: no cover
+        return False
+
+
+def prepare_smooth_shading(mesh, scalars, texture, split_sharp_edges, feature_angle):
+    """Prepare a dataset for smooth shading.
+
+    VTK requires datasets with prong shading to have active normals.
+    This requires extracting the external surfaces from non-polydata
+    datasets and computing the point normals.
+
+    Parameters
+    ----------
+    mesh : pyvista.DataSet
+        Dataset to prepare smooth shading for.
+    texture : vtk.vtkTexture or np.ndarray or bool, optional
+        A texture to apply to the mesh.
+    split_sharp_edges : bool, optional
+        Split sharp edges exceeding 30 degrees when plotting with
+        smooth shading.  Control the angle with the optional
+        keyword argument ``feature_angle``.  By default this is
+        ``False``.  Note that enabling this will create a copy of
+        the input mesh within the plotter.  See
+        :ref:`shading_example`.
+    feature_angle : float, optional
+        Angle to consider an edge a sharp edge.
+
+    Returns
+    -------
+    pyvista.PolyData
+        Always a surface as we need to compute point normals.
+
+    """
+    # extract surface if mesh is exterior
+    if not isinstance(mesh, pyvista.PolyData):
+        grid = mesh
+        mesh = grid.extract_surface()
+        # remap scalars
+        if scalars is not None:
+            ind = mesh.point_data['vtkOriginalPointIds']
+            scalars = np.asarray(scalars[ind])
```

Comment:
I thought about this for a bit, and I think there's ways we could make this marginally faster at the expense of complexity.

Since `extract_surface` always creates `vtkOriginalPointIds`, so we could simply use that array if the original dataset was a PolyData.  If not, we need to add the the `"__orig_ids__"` array to track indices.

I'll implement it as it shouldn't be too complex.

Reply:
Done.